oak_crypto/src/hpke.rs

```diff
@@ -0,0 +1,256 @@
+            .context("couldn't increment sequence number")?;
+        Ok(ciphertext)
+    }
+
+    /// Decrypts message and validates associated data using AEAD.
+    /// <https://www.rfc-editor.org/rfc/rfc9180.html#name-encryption-and-decryption>
+    pub(crate) fn open(
+        &mut self,
+        ciphertext: &AeadEncryptedMessage,
+    ) -> anyhow::Result<MessageWithAssociatedData> {
+        let nonce = self
+            .compute_nonce(self.sequence_number, &self.base_nonce)
+            .context("couldn't compute nonce")?;
+        let plaintext =
+            Aead::decrypt(&self.key, &nonce, ciphertext).context("couldn't decrypt message")?;
+        self.increment_sequence_number()
+            .context("couldn't increment sequence number")?;
+        Ok(plaintext)
+    }
+
+    /// Encrypts response message with associated data using AEAD as part of bidirectional
+    /// communication.
+    /// <https://www.rfc-editor.org/rfc/rfc9180.html#name-bidirectional-encryption>
+    pub(crate) fn seal_response(
+        &mut self,
+        plaintext: &MessageWithAssociatedData,
+    ) -> anyhow::Result<AeadEncryptedMessage> {
+        let nonce = self
+            .compute_nonce(self.sequence_number, &self.response_base_nonce)
+            .context("couldn't compute nonce")?;
+        let ciphertext = Aead::encrypt(&self.response_key, &nonce, plaintext)
+            .context("couldn't encrypt response message")?;
+        self.increment_sequence_number()
+            .context("couldn't increment sequence number")?;
+        Ok(ciphertext)
+    }
+
+    /// Decrypts response message and validates associated data using AEAD as part of bidirectional
+    /// communication.
+    /// <https://www.rfc-editor.org/rfc/rfc9180.html#name-bidirectional-encryption>
+    pub(crate) fn open_response(
+        &mut self,
+        ciphertext: &AeadEncryptedMessage,
+    ) -> anyhow::Result<MessageWithAssociatedData> {
+        let nonce = self
+            .compute_nonce(self.sequence_number, &self.response_base_nonce)
+            .context("couldn't compute nonce")?;
+        let plaintext = Aead::decrypt(&self.response_key, &nonce, ciphertext)
+            .context("couldn't decrypt response message")?;
+        self.increment_sequence_number()
+            .context("couldn't increment sequence number")?;
+        Ok(plaintext)
+    }
+
+    fn compute_nonce(
+        &self,
+        sequence_number: u128,
+        base_nonce: &AeadNonce,
+    ) -> anyhow::Result<AeadNonce> {
+        let sequence_number_bytes = i2osp::<AEAD_NONCE_SIZE_BYTES>(sequence_number)
+            .context("couldn't convert sequence number to bytes")?;
+        let nonce = xor::<AEAD_NONCE_SIZE_BYTES>(base_nonce, &sequence_number_bytes)
+            .context("couldn't xor base nonce and sequence number")?;
+        Ok(nonce)
+    }
+
+    fn increment_sequence_number(&mut self) -> anyhow::Result<()> {
+        if self.sequence_number >= ((1 << (8 * AEAD_NONCE_SIZE_BYTES)) - 1) {
```

Comment:
Maybe define this value as a const and explain how it's calculated.

Reply:
Done